### PR TITLE
Fix #33451: Refactor ai_chat component for iOS

### PR DESCRIPTION
**Pull request by Brandon-T**
## Summary
- Refactor `ai_chat_tab_helper` to be platform generic, so that iOS can use the same core code for its APIs.
- With this change, iOS is able to supply a web-page, URL, and all of the things Desktop does.
- iOS is now able to use the feedback API (shared code cannot be in `browser` or `content` as it's a layering violation, so this PR fixes that: https://www.chromium.org/developers/design-documents/layered-components-design/).

Resolves https://github.com/brave/brave-browser/issues/33451

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/ai_chat/core/ai_chat_helper.h [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    ~Observer() override {}
+
+    virtual void OnHistoryUpdate() {}
+    virtual void OnAPIRequestInProgress(bool in_progress) {}
+    virtual void OnAPIResponseError(ai_chat::mojom::APIError error) {}
+    virtual void OnSuggestedQuestionsChanged(
+        std::vector<std::string> questions,
+        bool has_generated,
+        ai_chat::mojom::AutoGenerateQuestionsPref auto_generate) {}
+    virtual void OnFaviconImageDataChanged() {}
```
Comment (sangwoo108) [inline components/ai_chat/core/ai_chat_helper.h:53]: thinking out loud - Favicon related callback might be moved else where, as this AIChatHelper doesn't handle it. Should we extend this observer from AIChatTabHelper?

**Review by sangwoo108 — APPROVED** [review]
Nice refactoring 👍 LGTM

File: components/ai_chat/core/ai_chat_helper.h (hunk 1 above)
Comment (Brandon-T) [reply components/ai_chat/core/ai_chat_helper.h:53]: Yeah correct. I might move it in another PR that I am working on atm where iOS gets fully integrated into Brave-Core :)
It should be handled by the TabHelper.

File: components/ai_chat/core/ai_chat_helper.h [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+#include <string>
+#include <vector>
+
+#include "base/memory/raw_ptr.h"
+#include "base/memory/ref_counted.h"
+#include "base/observer_list.h"
+#include "brave/components/ai_chat/common/mojom/ai_chat.mojom.h"
+#include "brave/components/ai_chat/core/engine/engine_consumer.h"
+#include "components/prefs/pref_change_registrar.h"
+#include "services/data_decoder/public/cpp/data_decoder.h"
```
Comment (mkarolin) [inline components/ai_chat/core/ai_chat_helper.h:19]: Is this needed?

File: components/ai_chat/core/ai_chat_helper.h [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    virtual void OnAPIResponseError(ai_chat::mojom::APIError error) {}
+    virtual void OnSuggestedQuestionsChanged(
+        std::vector<std::string> questions,
+        bool has_generated,
+        ai_chat::mojom::AutoGenerateQuestionsPref auto_generate) {}
+    virtual void OnFaviconImageDataChanged() {}
+    virtual void OnPageHasContent() {}
+  };
+
+  explicit AIChatHelper(AIChatHelperDelegate* delegate);
```
Comment (bridiver) [inline components/ai_chat/core/ai_chat_helper.h:57]: Please add comments explaining what this class does

Comment (bridiver) [reply components/ai_chat/core/ai_chat_helper.h:57]: also `tab_helper` is a chromium naming convention, but I'm not sure that `AIChatHelper` is the right name for this. It seems like it has a lot of responsibilities and I'm not sure it makes sense as a single class.

Comment (bridiver) [reply components/ai_chat/core/ai_chat_helper.h:57]: I'm not opposed to this as an intermediate refactoring with a follow-up to clean this up, but it definitely needs some comments

File: components/ai_chat/core/ai_chat_helper.cc [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+  chat_history_.clear();
+  engine_->ClearAllQueries();
+
+  // Trigger an observer update to refresh the UI.
+  for (auto& obs : observers_) {
+    obs.OnHistoryUpdate();
+  }
+}
+
+mojom::APIError AIChatHelper::GetCurrentAPIError() {
```
Comment (bridiver) [inline components/ai_chat/core/ai_chat_helper.cc:371]: This is not a great pattern and should be addressed in a follow-up. API methods should have callbacks to handle errors. Checking this is very unreliable, especially if you have more than one api call happening at the same time.

Comment (bridiver) [reply components/ai_chat/core/ai_chat_helper.cc:371]: commented on the original PR https://github.com/brave/brave-core/pull/19804/files#r1349226241

File: components/ai_chat/core/ai_chat_helper.h (hunk 3 above)
Comment (Brandon-T) [reply components/ai_chat/core/ai_chat_helper.h]: Renamed to `AIChatDriver` and refactored the code to get rid of the delegate and follow the `Driver` style.
So TabHelper now implements the driver and iOS will do the same.

File: components/ai_chat/core/ai_chat_helper.cc (hunk 4 above)
Comment (petemill) [reply components/ai_chat/core/ai_chat_helper.cc]: I responded over there, I think it's fine if not perfect.

File: components/ai_chat/core/ai_chat_helper.h (hunk 3 above)
Comment (petemill) [reply components/ai_chat/core/ai_chat_helper.h]: Didn't see this conversation when I just posted my other messages about this so adding here that I think naming this `Conversation` and using a Delegate is the correct approach. If not now, then very soon as follow-up or once we land support for standalone `Conversation`s which are not tied to a Tab / Content. The plan for that was to use a delegate to fetch an associated Tab for a Conversation. There would only be an associated Tab if 1) the conversation was started from the sidepanel of a Tab and 2) the conversation was not loaded from persistant storage.

File: components/ai_chat/core/ai_chat_helper.h (hunk 2 above)
Comment (Brandon-T) [reply components/ai_chat/core/ai_chat_helper.h]: I moved these and refactored the file completely so it is needed now.

**Review by petemill — APPROVED** [review]
Nicely done. Thanks for your patience following-through with many rebases.

Future follow-up suggestions (mostly with https://github.com/brave/brave-browser/issues/31623):
- Remove subclassing of Conversation when they can be standalone without an associated Tab.
- Provide potential page data to any conversation via a delegate with it's own per-platform implemenatation (e.g. ConversationContentDriver)
- Now that we've introduced core and made this a "layered component", everything needs to be in either core, content, iOS or resources. There shouldn't be a renderer subdirectory of ai_chat.
- Don't leak some implementation details from ConversationDriver to subclasses - e.g. subclasses have to know to call `CleanUp` and then `MaybeGeneratePageText` if they want to mark the start of a new "page".

**Review by kylehickinson — APPROVED** [review]

**Pull request merged**